From bf05ee0c2906f7db30ae5b56447c47e6cb48c9c2 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Tue, 24 Oct 2023 17:28:42 +0200
Subject: [PATCH 0595/1141] remoteproc: tee: Fix management of
 tee_shm_register_kernel_buf() fail

If the registration fails, return an appropriate error to prevent fw_shm
from becoming invalid. This fix addresses the issue where a kernel crash
occurs due to an invalid fw_shm.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Change-Id: I84ffd803cf27057c5f3e78d0db9fe2600361fa6d
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/335978
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/remoteproc/tee_remoteproc.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/remoteproc/tee_remoteproc.c b/drivers/remoteproc/tee_remoteproc.c
index bd4808105f4f..514080fdd38b 100644
--- a/drivers/remoteproc/tee_remoteproc.c
+++ b/drivers/remoteproc/tee_remoteproc.c
@@ -106,6 +106,11 @@ int tee_rproc_load_fw(struct tee_rproc *trproc, const struct firmware *fw)
 	int ret;
 
 	fw_shm = tee_shm_register_kernel_buf(pvt_data.ctx, (void *)fw->data, fw->size);
+	if (IS_ERR(fw_shm)) {
+		dev_err(pvt_data.dev, "Failed to register kernel buffer in tee: %ld\n",
+			PTR_ERR(fw_shm));
+		return PTR_ERR(fw_shm);
+	}
 
 	prepare_args(trproc, TA_RPROC_FW_CMD_LOAD_FW, &arg, param, 1);
 
-- 
2.39.2

